lib/oli_web/components/hierarchy_selector.ex

```diff
@@ -0,0 +1,162 @@
+          <div class="hierarchy-selector__selected-item">
+            {item_name}
+          </div>
+        {/for}
+      </div>
+      <div class="hierarchy-selector__list-container">
+        <select hidden name={"#{input_name(@form, @field)}[]"} multiple>
+          {#for {_item_name, item_id} <- @selected_values}
+            <option selected value={item_id} />
+          {/for}
+        </select>
+        <div data-active={"#{@expanded}"} class="hierarchy-selector__list" phx-update="ignore">
+          {#for item <- @items}
+            {render_resource_option(assigns, item, 0)}
+          {/for}
+        </div>
+        <div class="bg-transparent h-5" />
+      </div>
+    </div>
+    """
+  end
+
+  defp render_resource_option(assigns, item, level) do
+    ~F"""
+    <div
+      class="hierarchy-selector__item"
+      id={"hierarchy-selector__item-#{item.id}"}
+      data-expanded="false"
+      style={"margin-left: #{6 * level}px"}
+    >
+      <div class="flex gap-2 items-center">
+        <button
+          class={"hierarchy-selector__expand-button #{if length(item.children) == 0, do: "hidden"}"}
+          type="button"
+          onclick={"expandElement('hierarchy-selector__item-#{item.id}')"}
+        >
+          <i class="fa-solid fa-caret-down h-4 w-4" data-icon="caret-down" />
+          <i class="fa-solid fa-caret-up h-4 w-4" data-icon="caret-up" />
+        </button>
+        <Checkbox
+          value={Enum.member?(@selected_values, {item.name, "#{item.id}"})}
+          opts={["phx-target": @myself, "phx-value-item": "#{item.name}-#{item.id}"]}
+          click="select"
+          class="form-check-input"
+        />
+        <label>{item.name}</label>
+      </div>
+      <div class="hierarchy-selector__item-children">
+        {#for children_item <- item.children}
+          {render_resource_option(assigns, children_item, level + 1)}
+        {/for}
+      </div>
+    </div>
+    """
+  end
+
+  def handle_event("expand", _params, socket) do
+    if socket.assigns.expanded do
+      socket = assign(socket, expanded: false)
+      {:noreply, socket}
+    else
+      socket = assign(socket, expanded: true)
+      {:noreply, socket}
+    end
+  end
+
+  def handle_event("select", %{"item" => item} = params, socket) do
+    [_, item_name, item_id] = Regex.run(~r"(.*)-(\d+)", item)
+    checked = Map.get(params, "value", "false")
+
+    if checked === "true" do
+      selected_values = socket.assigns.selected_values ++ [{item_name, item_id}]
+
+      socket =
+        assign(
+          socket,
+          selected_values: selected_values
+        )
+
+      {:noreply, socket}
+    else
+      selected_values =
+        Enum.reject(socket.assigns.selected_values, fn {_, id} -> item_id == id end)
+
+      socket =
+        assign(
+          socket,
+          selected_values: selected_values
+        )
+
+      {:noreply, socket}
+    end
+  end
+
+  defp input_name(form_or_name, field)
```

Comment:
Just to understand... I'm not sure what this function returns here.

Reply:
Nothing really, I just stole this from `Phoenix.HTML.Form` 😅